File: ttshare_mobile/lib/services/processing_pipeline.dart
import 'dart:convert';
import 'package:path_provider/path_provider.dart';
import 'package:uuid/uuid.dart';
import 'package:intl/intl.dart';
import 'cookie_manager.dart';
import 'web_view_loader.dart';
import 'html_archiver.dart';
import 'readability_extractor.dart';
import 'webdav_client.dart';
import '../models/article_record.dart';

class ProcessResult {
  final ArticleRecord record;
  final bool success;
  final String? error;

  ProcessResult({required this.record, required this.success, this.error});
}

class ProcessingPipeline {
  final CookieManager _cookieManager;
  final WebViewLoader _webViewLoader;
  final HtmlArchiver _htmlArchiver;
  final ReadabilityExtractor _readabilityExtractor;
  final WebdavClient _webdavClient;

  ProcessingPipeline({
    required CookieManager cookieManager,
    required WebViewLoader webViewLoader,
    required HtmlArchiver htmlArchiver,
    required ReadabilityExtractor readabilityExtractor,
    required WebdavClient webdavClient,
  })  : _cookieManager = cookieManager,
        _webViewLoader = webViewLoader,
        _htmlArchiver = htmlArchiver,
        _readabilityExtractor = readabilityExtractor,
        _webdavClient = webdavClient;

  Future<ProcessResult> process(String url, String title) async {
    final id = const Uuid().v4();
    final dateStr = DateFormat('yyyy-MM-dd').format(DateTime.now());
    final source = Uri.tryParse(url)?.host ?? 'unknown';
    final folderName = '$dateStr-$source-${_sanitize(title)}';
    final safeFolder = folderName.substring(0, folderName.length.clamp(1, 100));

    try {
      // 1. Check cookies and load page via headless WebView
      final domain = Uri.parse(url).host;
      final cookieEntry = await _cookieManager.getCookies(domain);
      final html = await _webViewLoader.loadPage(
        url: url,
        cookieData: cookieEntry?.cookieData,
      );

      // 2. Create HTML archive
      final htmlPath = await _htmlArchiver.archive(html, title);

      // 3. Extract Markdown via Readability
      final extractResult = await _readabilityExtractor.extract(html, url);
      String? mdLocalPath;

      final rawJson = extractResult['raw'] as String?;
      if (rawJson != null) {
        try {
          final parsed = json.decode(rawJson) as Map<String, dynamic>;
          final markdown = parsed['markdown'] as String?;
          if (markdown != null && markdown.isNotEmpty) {
            // Save markdown file alongside HTML
            final tmpDir = await getTemporaryDirectory();
            final mdDir = '${tmpDir.path}/${_sanitize(title)}';
            mdLocalPath = '$mdDir/index.md';
          }
        } catch (_) {}
      }

      // 4. Upload to WebDAV
      if (!_webdavClient.isConfigured) {
        return ProcessResult(
          record: ArticleRecord(
            id: id,
            title: title,
            source: source,
            url: url,
            savedAt: DateTime.now(),
            status: 'failed',
            errorMessage: 'WebDAV not configured',
          ),
          success: false,
          error: 'WebDAV not configured',
        );
      }

      final basePath = '/$safeFolder';
      await _webdavClient.createFolder(basePath);

      final htmlRemotePath = '$basePath/index.html';
      await _webdavClient.uploadFile(htmlPath, htmlRemotePath);

      if (mdLocalPath != null) {
        final mdRemotePath = '$basePath/index.md';
        await _webdavClient.uploadFile(mdLocalPath, mdRemotePath);
      }

      return ProcessResult(
        record: ArticleRecord(
          id: id,
          title: title,
          source: source,
          url: url,
          savedAt: DateTime.now(),
          status: 'completed',
          htmlWebdavPath: htmlRemotePath,
        ),
        success: true,
      );
    } catch (e) {
      return ProcessResult(
        record: ArticleRecord(
          id: id,
          title: title,
          source: source,
          url: url,
          savedAt: DateTime.now(),
          status: 'failed',
          errorMessage: e.toString(),
        ),
        success: false,
        error: e.toString(),
      );
    }
  }

  String _sanitize(String s) {
    return s
        .replaceAll(RegExp(r'[<>:"/\\|?*]'), '_')
        .replaceAll(RegExp(r'\s+'), '_')
        .trim();
  }
}

File: ttshare_android/app/src/main/java/com/ttshare/service/ProcessingPipeline.kt
package com.ttshare.service

import android.content.Context
import com.ttshare.data.ArticleRecord
import com.ttshare.data.AppDatabase
import java.io.File
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*

class ProcessingPipeline(
    private val context: Context,
    private val prefs: android.content.SharedPreferences
) {
    private val webViewLoader = WebViewLoader()
    private val htmlArchiver = HtmlArchiver()
    private val readabilityExtractor = ReadabilityExtractor()
    private val webdavClient = WebdavClient(prefs)
    private val cookieManager = CookieManager(prefs)
    private val db = AppDatabase.getInstance(context)

    data class ProcessResult(
        val record: ArticleRecord,
        val success: Boolean,
        val error: String? = null
    )

    suspend fun process(url: String, title: String): ProcessResult {
        val id = UUID.randomUUID().toString()
        val dateStr = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date())
        val source = try { URL(url).host } catch (_: Exception) { "unknown" }
        val folderName = "$dateStr-$source-${htmlArchiver.sanitizeFileName(title)}".take(100)

        return try {
            // 1. Load page with cookies
            val domain = URL(url).host
            val cookies = cookieManager.getCookies(domain)
            val html = webViewLoader.loadPage(url = url, cookieData = cookies?.cookieData)

            // 2. Create temp directory
            val cacheDir = File(context.cacheDir, folderName)
            cacheDir.mkdirs()

            // 3. Archive HTML
            val htmlPath = htmlArchiver.archive(html, title, cacheDir)

            // 4. Extract markdown
            val extractResult = readabilityExtractor.extract(html)
            var mdPath: String? = null
            if (extractResult.markdown != null) {
                mdPath = htmlArchiver.saveMarkdown(extractResult.markdown, cacheDir)
            }

            // 5. Upload to WebDAV
            if (!webdavClient.isConfigured) {
                throw Exception("WebDAV not configured")
            }
            webdavClient.createFolder("/$folderName")
            val htmlRemote = "/$folderName/index.html"
            webdavClient.uploadFile(htmlPath, htmlRemote)

            if (mdPath != null) {
                webdavClient.uploadFile(mdPath, "/$folderName/index.md")
            }

            val record = ArticleRecord(
                id = id,
                title = title,
                source = source,
                url = url,
                savedAt = System.currentTimeMillis(),
                status = "completed",
                htmlWebdavPath = htmlRemote
            )

            db.articleDao().upsert(record)

            ProcessResult(record, true)
        } catch (e: Exception) {
            val record = ArticleRecord(
                id = id,
                title = title,
                source = source,
                url = url,
                savedAt = System.currentTimeMillis(),
                status = "failed",
                errorMessage = e.message
            )
            db.articleDao().upsert(record)
            ProcessResult(record, false, e.message)
        }
    }
}
